src/content/__tests__/monetizationLinkManager.test.ts

```diff
@@ -0,0 +1,1236 @@
+import { JSDOM } from 'jsdom';
+import { MonetizationLinkManager } from '@/content/services/monetizationLinkManager';
+import type {
+  ContentToBackgroundMessage,
+  MessageManager,
+  Response,
+} from '@/shared/messages';
+import type { Logger } from '@/shared/logger';
+import type { WalletAddress } from '@interledger/open-payments';
+import { success, failure } from '@/shared/helpers';
+
+// for syntax highlighting
+const html = String.raw;
+
+const WALLET_ADDRESS: WalletAddress[] = [
+  {
+    id: '$wallet.example.com/alice',
+    authServer: 'https://auth.example.com',
+    resourceServer: 'https://wallet.example.com',
+    publicName: 'Alice USD Wallet',
+    assetCode: 'USD',
+    assetScale: 2,
+  },
+  {
+    id: '$wallet.example.com/bob',
+    authServer: 'https://auth2.example.com',
+    resourceServer: 'https://wallet2.example.com',
+    publicName: 'Bob EUR Wallet',
+    assetCode: 'EUR',
+    assetScale: 2,
+  },
+  {
+    id: '$wallet.example.com/carol',
+    authServer: 'https://auth3.example.com',
+    resourceServer: 'https://wallet3.example.com',
+    publicName: 'Carol MXN Wallet',
+    assetCode: 'MXN',
+    assetScale: 2,
+  },
+];
+
+describe('MonetizationLinkManager', () => {
+  const messageManager = {
+    send: () => {},
+  } as unknown as MessageManager<ContentToBackgroundMessage>;
+  const loggerMock = {
+    error: jest.fn(),
+  } as unknown as Logger;
+  const requestIdMock = jest.fn(() => `request-${Math.random()}`);
+
+  const msg: {
+    [k in keyof ContentToBackgroundMessage]: jest.Mock<
+      Promise<Response<ContentToBackgroundMessage[k]['output']>>,
+      [ContentToBackgroundMessage[k]['input']]
+    >;
+  } = {
+    GET_WALLET_ADDRESS_INFO: jest.fn(),
+    RESUME_MONETIZATION: jest.fn(),
+    START_MONETIZATION: jest.fn(),
+    STOP_MONETIZATION: jest.fn(),
+    TAB_FOCUSED: jest.fn(),
+  };
+  const messageMock = jest.spyOn(messageManager, 'send');
+  // @ts-expect-error let it go
+  messageMock.mockImplementation((action, payload) => msg[action](payload));
+
+  function createMonetizationLinkManager(document: Document) {
+    const linkManager = new MonetizationLinkManager({
+      global: document.defaultView!.globalThis,
+      document: document,
+      message: messageManager,
+      logger: loggerMock,
+    });
+
+    return linkManager;
+  }
+
+  function createTestEnv({
+    head = '',
+    body = '',
+  }: { head?: string; body?: string }) {
+    const htm = html`<!DOCTYPE html><html><head>${head}</head><body>${body}</body></html>`;
+    const dom = new JSDOM(htm, {
+      runScripts: 'dangerously',
+      pretendToBeVisual: true,
+      resources: 'usable',
+    });
+
+    const window = dom.window;
+    const document = window.document;
+
+    return {
+      dom,
+      window,
+      document,
+      documentReadyState: jest.spyOn(document, 'readyState', 'get'),
+    };
+  }
```

Comment:
Let's avoid repeating ourselves.

```suggestion
  function createTestEnv({
    head = '',
    body = '',
    interactive = false,
  }: { head?: string; body?: string; interactive?: boolean }) {
    const htm = html`<!DOCTYPE html><html><head>${head}</head><body>${body}</body></html>`;
    const dom = new JSDOM(htm, {
      runScripts: 'dangerously',
      pretendToBeVisual: true,
      resources: 'usable',
    });

    const window = dom.window;
    const document = window.document;

    const documentReadyState = jest.spyOn(document, 'readyState', 'get');
    if (interactive) {
      documentReadyState.mockReturnValue('interactive');
    }

    return {
      dom,
      window,
      document,
      documentReadyState,
    };
  }
```

Reply:
Or maybe can pass `readyState = "interactive"` by default (as all current tests do that only), and then can override it manually later when needed. Like pass `{readyState:null}` to avoid setting it in `createTestEnv` and then later set it manually via `documentReadyState`.